Lib/http/server.py

```diff
@@ -881,6 +878,23 @@ def guess_type(self, path):
         return 'application/octet-stream'
 
 
+def _get_redirect_url(path):
+    """Returns URL with trailing slash on path, if required.  If not required,
+    returns None.
+    """
+    # Previous versions of this module used urllib.parse.urlsplit() here.
+    # However, the 'path' is not truly a URI in that it can't have a scheme or
+    # netloc.  We need to avoid parsing it incorrectly.  For example, as
+    # reported in gh-87389, a path starting with a double slash should not be
+    # treated as a relative URI.  Also, a path with a colon in the first
+    # component could also be parsed wrongly.
+    parts = urllib.parse.pathsplit(path)
```

Comment:
Unfortunately BaseHTTPRequestHandler `self.path` isn't guaranteed to be only a pathname by anything.

`GET http://netloc/path/to/thing HTTP/1.0`

Is a valid HTTP query to many http servers.  Our existing http server code ignores the scheme and netloc on that and serves up /path/to/thing properly.  This code _might_ break things doing that?  ie: if http://netloc winds up in the returned path, the it seems like that'd move through pathsplit and urlunsplit properly only by chance rather than design just given the name and purpose of the pathsplit API vs what was just handed to it as input.

Check out my updated test in https://github.com/python/cpython/issues/87389.

I have no idea if anything depends on this as it should be unusual for any client not pretending to be talking to an HTTP proxy to make requests that way as that isn't normal per the http/1.0 and 1.1 RFCs but as it worked before and other servers like Apache continue to support it, I don't think we can break that - at least not in a bugfix.

Reply:
Based on my reading of the RFC, `GET http://netloc/path/to/thing HTTP/1.0` should technically be allowed.  However, `http.server` doesn't handle it, before or after this PR.  Instead, it tries to use the second word as an absolute file path. I feel like we should not attempt to fix `http.server` to handle the above.  For HTTP 1.0, using only a file path was required.  For `curl` I notice that even with the `--http1.1` flag, it passes a path and not the URL with the scheme+netloc.

The bug causing #87389, IMHO, is that `http.server` was not consistent in how it was treating `self.path`.  The `translate_path()` method was treating it as a path and not a URL.  But, the redirect to add a trailing slash (done only if a folder with that name was found), parsed it as allowing a full URL.  It needs to do it one way or the other consistently.  The `_request_path_split()` method I added does this.  It doesn't make sense to add this to `urllib.parse` since what `http.server` is doing is not modern HTTP.